### ISSUE TITLE
terraform crashing on validate

### ISSUE THREAD

**Issue by thexder1** [hashicorp/terraform#21733]
### Terraform Version

```
terraform --version
Terraform v0.12.2
+ provider.google v2.8.0

...
```

### Terraform Configuration Files

tf file below
```hcl
variable "zone" {
    type = string
}
variable "names" {
    type = list(string)
}
variable "records" {
    type = "map"
}
variable "ttl" {
    type = number
}

resource "google_dns_managed_zone" "prod" {
    name = var.zone
    dns_name = var.zone
}

resource "google_dns_record_set" "prod" {
    dynamic "name" {
        for_each = var.names
        content {
            dynamic "record" {
                for_each = var.records[name.value]
                content {
                    name = name.value
                    type = record.key
                    ttl = var.ttl
                    managed_zone = google_dns_managed_zone.prod.name
                    rrdatas = record.value[*]
                }
            }
        }
    }
}

...
```
tfvars file below
```hcl
ttl = 1800
records = {
    mydomain.com. = {
        "MX" = [(0,"mxmail.mydomain.com.")]
        "A" = [127.0.0.1]
        "NS" = [ns1.mydomain.com,ns2.mydomain.com]
    }
    *.mydomain.com. = {
        "MX" = [(0,"mxmail.mydomain.com.")]
        "A" = [127.0.0.1]
    }
}
names = [mydomain.com, *.mydomain.com]


...
```

### Debug Output

```
2019/06/14 11:45:10 [INFO] Terraform version: 0.12.2  
2019/06/14 11:45:10 [INFO] Go runtime version: go1.12.4
2019/06/14 11:45:10 [INFO] CLI args: []string{"/usr/bin/terraform", "validate", "-var-file=test.tfvars"}
2019/06/14 11:45:10 [DEBUG] Attempting to open CLI config file: /home/jparker/.terraformrc
2019/06/14 11:45:10 [DEBUG] File doesn't exist, but doesn't need to. Ignoring.
2019/06/14 11:45:10 [INFO] CLI command args: []string{"validate", "-var-file=test.tfvars"}
2019/06/14 11:45:10 [DEBUG] checking for provider in "."
2019/06/14 11:45:10 [DEBUG] checking for provider in "/usr/bin"
2019/06/14 11:45:10 [DEBUG] checking for provider in ".terraform/plugins/linux_amd64"
2019/06/14 11:45:10 [DEBUG] found provider "terraform-provider-google_v2.8.0_x4"
2019/06/14 11:45:10 [DEBUG] found valid plugin: "google", "2.8.0", "/home/jparker/repos/terraform/build/zones/.terraform/plugins/linux_amd64/terraform-provider-google_v2.8.0_x4"
2019/06/14 11:45:10 [DEBUG] checking for provisioner in "."
2019/06/14 11:45:10 [DEBUG] checking for provisioner in "/usr/bin"
2019/06/14 11:45:10 [DEBUG] checking for provisioner in ".terraform/plugins/linux_amd64"
2019/06/14 11:45:10 [TRACE] terraform.NewContext: starting
2019/06/14 11:45:10 [TRACE] terraform.NewContext: resolving provider version selections
2019/06/14 11:45:11 [TRACE] terraform.NewContext: loading provider schemas
2019/06/14 11:45:11 [TRACE] LoadSchemas: retrieving schema for provider type "google"
2019-06-14T11:45:11.120-0500 [INFO]  plugin: configuring client automatic mTLS
2019-06-14T11:45:11.155-0500 [DEBUG] plugin: starting plugin: path=/home/jparker/repos/terraform/build/zones/.terraform/plugins/linux_amd64/terraform-provider-google_v2.8.0_x4 args=[/home/jparker/repos/terraform/build/zones/.terraform/plugins/linux_amd64/terraform-provider-google_v2.8.0_x4]
2019-06-14T11:45:11.155-0500 [DEBUG] plugin: plugin started: path=/home/jparker/repos/terraform/build/zones/.terraform/plugins/linux_amd64/terraform-provider-google_v2.8.0_x4 pid=13910
2019-06-14T11:45:11.155-0500 [DEBUG] plugin: waiting for RPC address: path=/home/jparker/repos/terraform/build/zones/.terraform/plugins/linux_amd64/terraform-provider-google_v2.8.0_x4
2019-06-14T11:45:11.164-0500 [INFO]  plugin.terraform-provider-google_v2.8.0_x4: configuring server automatic mTLS: timestamp=2019-06-14T11:45:11.164-0500
2019-06-14T11:45:11.200-0500 [DEBUG] plugin.terraform-provider-google_v2.8.0_x4: plugin address: address=/tmp/plugin383768069 network=unix timestamp=2019-06-14T11:45:11.200-0500
2019-06-14T11:45:11.200-0500 [DEBUG] plugin: using plugin: version=5
2019/06/14 11:45:11 [TRACE] GRPCProvider: GetSchema
2019/06/14 11:45:11 [TRACE] GRPCProvider: Close
2019-06-14T11:45:11.297-0500 [DEBUG] plugin: plugin process exited: path=/home/jparker/repos/terraform/build/zones/.terraform/plugins/linux_amd64/terraform-provider-google_v2.8.0_x4 pid=13910
2019-06-14T11:45:11.297-0500 [DEBUG] plugin: plugin exited
2019/06/14 11:45:11 [TRACE] terraform.NewContext: complete
2019/06/14 11:45:11 [INFO] terraform: building graph: GraphTypeValidate
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.ConfigTransformer
2019/06/14 11:45:11 [TRACE] ConfigTransformer: Starting for path: 
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.ConfigTransformer with new graph:
google_dns_managed_zone.prod - *terraform.NodeValidatableResource
google_dns_record_set.prod - *terraform.NodeValidatableResource
------
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.LocalTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.LocalTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.OutputTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.OutputTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.OrphanResourceInstanceTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.OrphanResourceInstanceTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.StateTransformer
2019/06/14 11:45:11 [TRACE] StateTransformer: state is empty, so nothing to do
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.StateTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.OrphanOutputTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.OrphanOutputTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.AttachResourceConfigTransformer
2019/06/14 11:45:11 [TRACE] AttachResourceConfigTransformer: attaching to "google_dns_managed_zone.prod" (*terraform.NodeValidatableResource) config from zones.tf:14,1-42
2019/06/14 11:45:11 [TRACE] AttachResourceConfigTransformer: attaching to "google_dns_record_set.prod" (*terraform.NodeValidatableResource) config from zones.tf:19,1-40
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.AttachResourceConfigTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.AttachStateTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.AttachStateTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.RootVariableTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.RootVariableTransformer with new graph:
google_dns_managed_zone.prod - *terraform.NodeValidatableResource
google_dns_record_set.prod - *terraform.NodeValidatableResource
var.names - *terraform.NodeRootVariable
var.records - *terraform.NodeRootVariable
var.ttl - *terraform.NodeRootVariable
var.zone - *terraform.NodeRootVariable
------
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.MissingProvisionerTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.MissingProvisionerTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.ProvisionerTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.ProvisionerTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.ModuleVariableTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.ModuleVariableTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.graphTransformerMulti
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderConfigTransformer
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderConfigTransformer with new graph:
google_dns_managed_zone.prod - *terraform.NodeValidatableResource
google_dns_record_set.prod - *terraform.NodeValidatableResource
var.names - *terraform.NodeRootVariable
var.records - *terraform.NodeRootVariable
var.ttl - *terraform.NodeRootVariable
var.zone - *terraform.NodeRootVariable
------
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.MissingProviderTransformer
2019/06/14 11:45:11 [DEBUG] adding implicit provider configuration provider.google, implied first by google_dns_managed_zone.prod
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.MissingProviderTransformer with new graph:
google_dns_managed_zone.prod - *terraform.NodeValidatableResource
google_dns_record_set.prod - *terraform.NodeValidatableResource
provider.google - *terraform.NodeApplyableProvider
var.names - *terraform.NodeRootVariable
var.records - *terraform.NodeRootVariable
var.ttl - *terraform.NodeRootVariable
var.zone - *terraform.NodeRootVariable
------
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ProviderTransformer
2019/06/14 11:45:11 [TRACE] ProviderTransformer: google_dns_managed_zone.prod is provided by provider.google or inherited equivalent
2019/06/14 11:45:11 [TRACE] ProviderTransformer: google_dns_record_set.prod is provided by provider.google or inherited equivalent
2019/06/14 11:45:11 [TRACE] ProviderTransformer: exact match for provider.google serving google_dns_record_set.prod
2019/06/14 11:45:11 [DEBUG] ProviderTransformer: "google_dns_record_set.prod" (*terraform.NodeValidatableResource) needs provider.google
2019/06/14 11:45:11 [TRACE] ProviderTransformer: exact match for provider.google serving google_dns_managed_zone.prod
2019/06/14 11:45:11 [DEBUG] ProviderTransformer: "google_dns_managed_zone.prod" (*terraform.NodeValidatableResource) needs provider.google
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ProviderTransformer with new graph:
google_dns_managed_zone.prod - *terraform.NodeValidatableResource
  provider.google - *terraform.NodeApplyableProvider
google_dns_record_set.prod - *terraform.NodeValidatableResource
  provider.google - *terraform.NodeApplyableProvider
provider.google - *terraform.NodeApplyableProvider
var.names - *terraform.NodeRootVariable
var.records - *terraform.NodeRootVariable
var.ttl - *terraform.NodeRootVariable
var.zone - *terraform.NodeRootVariable
------
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.PruneProviderTransformer
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.PruneProviderTransformer (no changes)
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Executing graph transform *terraform.ParentProviderTransformer
2019/06/14 11:45:11 [TRACE] (graphTransformerMulti) Completed graph transform *terraform.ParentProviderTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.graphTransformerMulti with new graph:
google_dns_managed_zone.prod - *terraform.NodeValidatableResource
  provider.google - *terraform.NodeApplyableProvider
google_dns_record_set.prod - *terraform.NodeValidatableResource
  provider.google - *terraform.NodeApplyableProvider
provider.google - *terraform.NodeApplyableProvider
var.names - *terraform.NodeRootVariable
var.records - *terraform.NodeRootVariable
var.ttl - *terraform.NodeRootVariable
var.zone - *terraform.NodeRootVariable
------
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.RemovedModuleTransformer
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.RemovedModuleTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.AttachSchemaTransformer
2019/06/14 11:45:11 [TRACE] AttachSchemaTransformer: attaching provider config schema to provider.google
2019/06/14 11:45:11 [TRACE] AttachSchemaTransformer: attaching resource schema to google_dns_managed_zone.prod
2019/06/14 11:45:11 [TRACE] AttachSchemaTransformer: attaching resource schema to google_dns_record_set.prod
2019/06/14 11:45:11 [TRACE] Completed graph transform *terraform.AttachSchemaTransformer (no changes)
2019/06/14 11:45:11 [TRACE] Executing graph transform *terraform.ReferenceTransformer
panic: not a collection type

goroutine 1 [running]:
github.com/zclconf/go-cty/cty.Type.ElementType(0x21eabe0, 0xc000040469, 0xc000041020, 0x4)
        /opt/gopath/pkg/mod/github.com/zclconf/go-cty@v0.0.0-20190516203816-4fecf87372ec/cty/collection.go:25 +0xd2
github.com/hashicorp/terraform/lang/blocktoattr.walkVariables(0x21e9860, 0xc0002fd550, 0x0, 0xc00033f201, 0x21e9860, 0xc0002fd550, 0xc000319de0, 0x40b849, 0x73, 0xc00015afac)
        /opt/gopath/src/github.com/hashicorp/terraform/lang/blocktoattr/variables.go:37 +0x45e
github.com/hashicorp/terraform/lang/blocktoattr.ExpandedVariables(...)
        /opt/gopath/src/github.com/hashicorp/terraform/lang/blocktoattr/variables.go:24
github.com/hashicorp/terraform/lang.ReferencesInBlock(0x21e9860, 0xc0002fd550, 0xc000319de0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x7f3ffd1cd008)
        /opt/gopath/src/github.com/hashicorp/terraform/lang/references.go:68 +0x66
github.com/hashicorp/terraform/terraform.(*NodeAbstractResource).References(0xc000483ce0, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/node_resource_abstract.go:190 +0x251
github.com/hashicorp/terraform/terraform.NewReferenceMap(0xc000113b20, 0x7, 0x7, 0x7)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/transform_reference.go:402 +0x24a
github.com/hashicorp/terraform/terraform.(*ReferenceTransformer).Transform(0x32e0c40, 0xc0000bf630, 0x14, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/transform_reference.go:73 +0x71
github.com/hashicorp/terraform/terraform.(*BasicGraphBuilder).Build(0xc00033f798, 0x0, 0x0, 0x0, 0x32c0f80, 0xc00033f7c8, 0x4e5650, 0xc0000be000)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/graph_builder.go:50 +0x477
github.com/hashicorp/terraform/terraform.(*PlanGraphBuilder).Build(0xc0001127e0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x1a6f440, 0xc0004078c0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/graph_builder_plan.go:65 +0xbc
github.com/hashicorp/terraform/terraform.(*Context).Graph(0xc0000180f0, 0x6, 0xc00033f81e, 0xc00033f900, 0x1, 0x1, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/context.go:292 +0x1b8
github.com/hashicorp/terraform/terraform.(*Context).Validate(0xc0000180f0, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/context.go:670 +0xf3
github.com/hashicorp/terraform/command.(*ValidateCommand).validate(0xc0003ef500, 0xc0003fc150, 0x29, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/command/validate.go:114 +0x4f9
github.com/hashicorp/terraform/command.(*ValidateCommand).Run(0xc0003ef500, 0xc00003a1a0, 0x1, 0x1, 0xc000091b60)
        /opt/gopath/src/github.com/hashicorp/terraform/command/validate.go:70 +0x6c2
github.com/mitchellh/cli.(*CLI).Run(0xc0000fe640, 0xc0000fe640, 0xc000531d90, 0x1)
        /opt/gopath/pkg/mod/github.com/mitchellh/cli@v1.0.0/cli.go:255 +0x1f1
main.wrappedMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:223 +0xaff
main.realMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:100 +0xb4
main.main()
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:36 +0x3b
panic: not a collection type

goroutine 1 [running]:
github.com/zclconf/go-cty/cty.Type.ElementType(0x21eabe0, 0xc000040469, 0xc000041020, 0x4)
        /opt/gopath/pkg/mod/github.com/zclconf/go-cty@v0.0.0-20190516203816-4fecf87372ec/cty/collection.go:25 +0xd2
github.com/hashicorp/terraform/lang/blocktoattr.walkVariables(0x21e9860, 0xc0002fd550, 0x0, 0xc00033f201, 0x21e9860, 0xc0002fd550, 0xc000319de0, 0x40b849, 0x73, 0xc00015afac)
        /opt/gopath/src/github.com/hashicorp/terraform/lang/blocktoattr/variables.go:37 +0x45e
github.com/hashicorp/terraform/lang/blocktoattr.ExpandedVariables(...)
        /opt/gopath/src/github.com/hashicorp/terraform/lang/blocktoattr/variables.go:24
github.com/hashicorp/terraform/lang.ReferencesInBlock(0x21e9860, 0xc0002fd550, 0xc000319de0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x7f3ffd1cd008)
        /opt/gopath/src/github.com/hashicorp/terraform/lang/references.go:68 +0x66
github.com/hashicorp/terraform/terraform.(*NodeAbstractResource).References(0xc000483ce0, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/node_resource_abstract.go:190 +0x251
github.com/hashicorp/terraform/terraform.NewReferenceMap(0xc000113b20, 0x7, 0x7, 0x7)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/transform_reference.go:402 +0x24a
github.com/hashicorp/terraform/terraform.(*ReferenceTransformer).Transform(0x32e0c40, 0xc0000bf630, 0x14, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/transform_reference.go:73 +0x71
github.com/hashicorp/terraform/terraform.(*BasicGraphBuilder).Build(0xc00033f798, 0x0, 0x0, 0x0, 0x32c0f80, 0xc00033f7c8, 0x4e5650, 0xc0000be000)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/graph_builder.go:50 +0x477
github.com/hashicorp/terraform/terraform.(*PlanGraphBuilder).Build(0xc0001127e0, 0x0, 0x0, 0x0, 0x1, 0x0, 0x1a6f440, 0xc0004078c0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/graph_builder_plan.go:65 +0xbc
github.com/hashicorp/terraform/terraform.(*Context).Graph(0xc0000180f0, 0x6, 0xc00033f81e, 0xc00033f900, 0x1, 0x1, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/context.go:292 +0x1b8
github.com/hashicorp/terraform/terraform.(*Context).Validate(0xc0000180f0, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/terraform/context.go:670 +0xf3
github.com/hashicorp/terraform/command.(*ValidateCommand).validate(0xc0003ef500, 0xc0003fc150, 0x29, 0x0, 0x0, 0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/command/validate.go:114 +0x4f9
github.com/hashicorp/terraform/command.(*ValidateCommand).Run(0xc0003ef500, 0xc00003a1a0, 0x1, 0x1, 0xc000091b60)
        /opt/gopath/src/github.com/hashicorp/terraform/command/validate.go:70 +0x6c2
github.com/mitchellh/cli.(*CLI).Run(0xc0000fe640, 0xc0000fe640, 0xc000531d90, 0x1)
        /opt/gopath/pkg/mod/github.com/mitchellh/cli@v1.0.0/cli.go:255 +0x1f1
main.wrappedMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:223 +0xaff
main.realMain(0x0)
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:100 +0xb4
main.main()
        /opt/gopath/src/github.com/hashicorp/terraform/main.go:36 +0x3b
```

### Steps to Reproduce

Use provided files
terraform init
terraform validate

**Comment by mildwonkey** [comment]
Hi @thexder1, thanks for reporting this and I'm sorry you've experienced it.

While the crash should not happen, and that's something we will address, I believe it's happening because you are using a `dynamic` block with a resource that does not have any nested blocks. `dynamic` cannot be used to create multiple resources; it is used within a resource that has repeatable nested blocks. You can read more about the `dynamic` construct [here](https://www.terraform.io/docs/configuration/expressions.html#dynamic-blocks).

**Comment by thexder1** [comment]
Thanks for the info. I guess then I am not sure how to do what I am trying to do as I have a bunch of DNS zones to manage that can have arbitrary number of names with arbitrary record types for each name.

**Comment by apparentlymart** [comment]
Hi @thexder1,

There is an example similar to what you are trying to do here in a [Google Cloud DNS Recordsets Module](https://github.com/terraformdns/terraform-google-dns-recordsets). It uses the `count` meta-argument to create one recordset for each entry in a given list.

That technique has the important caveat that if you make any later changes to the given list then it will cause changes to any objects representing later recordsets in the list because Terraform will correlate them by their positions in the list rather than by any specific identifier inside the object. That limitation will be addressed by the feature being discussed in #17179, but careful use of `count` can be a good interim solution if you are able to tolerate the renumbering issue.

**Comment by x-fauzan-x** [comment]
Happened on me . Atom can't be opened after I tried to open terraform state file
